sdk/appconfiguration/AzureAppConfiguration/Source/AppConfigurationClientOptions.swift

```diff
@@ -29,10 +29,10 @@ import Foundation
 
 public class AppConfigurationClientOptions: AzureConfigurable {
     public let apiVersion: String
-    public let logger: ClientLogger
+    public let logger: ClientLoggerProtocol
```

Comment:
Not all protocols should. I've been trying to cobble together best practices from sources like https://github.com/raywenderlich/swift-style-guide and https://softwareengineering.stackexchange.com/questions/263521/swift-protocol-naming-conventions as well as observations from the Swift standard libraries. 

The guideline I used here is that I didn't want a protocol to be confused with a class. So, it was something like "ClientLoggable" or "ClientLogging", that would signal "protocol" to me. Also "ClientLoggerDelegate" if it is used as part of a delegate pattern. In this case, "ClientLogger" reads like a class, so to disambiguate, I added "Protocol", the simplest way of making it clear that it's a protocol and not a class.

Reply:
It would be good to nail down our own conventions rather than relying on various opinions on the web.

Reply:
I think "the thing you should be returning" should be "XXX" and "the implementation of the thing" should be "idXXXImpl".  e.g.

* `ClientLogger` the protocol
* `OSClientLoggerImpl` the implementation

Unless we want to start all implementations with "XXXImpl" and then after - good for sorting within the file browser

* `ClientLogger`
* `ClientLoggerImplOS`
* `ClientLoggerImplNull`

Thoughts?

Reply:
Here's an excerpt from the [Swift API design guidelines](https://swift.org/documentation/api-design-guidelines/):

```
Protocols that describe what something is should read as nouns (e.g. Collection).

Protocols that describe a capability should be named using the suffixes able, ible, or ing (e.g. Equatable, ProgressReporting).

...

If an associated type is so tightly bound to its protocol constraint that the protocol name is the role, avoid collision by appending Protocol to the protocol name:

protocol Sequence {
  associatedtype Iterator : IteratorProtocol
}
protocol IteratorProtocol { ... }
```

So based on this, I think I've overused the Protocol suffix and will go back and re-evaluate. I've not seen any precedent for Impl in iOS so far, so I think the old names were fine. I tried to standardize one way and it probably should have been the other.

So:
ClientLogger (Protocol)
ClientLoggerOS (class)
ClientLoggerNull (class)

I will have to convert OTHER cases where things end in Protocol to one of the other patterns.

Reply:
I like the Swift API design guidelines.  We should link those in our guidelines (except where otherwise stated in these guidelines, use the Swift API design guidelines")

Based on a quick read:

* `ClientLogger` (noun - the protocol)
* `OSClientLogging` (an implementation)
* `NullClientLogging` (also an implementation)

Makes sense to me and in line with official guidance from Apple.

Reply:
Based on the guidelines, it seems like both the protocol and the types that implement it should be nouns (emphasis mine):

> Protocols that describe what something is should read as nouns (e.g. Collection).
>
> Protocols that describe a capability should be named using the suffixes able, ible, or ing (e.g. Equatable, ProgressReporting).
>
> **The names of other types, properties, variables, and constants should read as nouns.**

So it sounds like the guidelines do explicitly allow for the possibility of confusion between the protocol and a type that conforms to that protocol. Though the IDE is still there to help disambiguate (I think Xcode stylizes protocols and classes/structs differently by default).,

Thus I'd propose perhaps `OSClientLogger` and `NullClientLogger` - what do you think?

Reply:
I agree that both the protocol and the class can be nouns provided that the protocol describes what something is (which is true for ClientLogger). 

I like OSClientLogger and NullClientLogger. Using an -ing ending or -able ending would suggest a protocol, not a class.

Reply:
So I might need your advice for some of the other protocol names :)

Reply:
Happy to contribute as needed, but also eager to avoid bikeshedding about names - if it sounds good and matches the guidelines feel free to just go with it :)

Reply:
I've managed to remove all `*Protocol` names by converting to a capability (-ing, -able) name or an "is" noun. 

```
Authenticating
AzureConfigurable
ClientLogger
Copyable
DataStringConvertible
HTTPTransportStage
MSALInteractiveDelegate
ObjectiveCBridgeable
PagedCollectionDelegate
PipelineContextSupporting
PipelineStage
TokenCredential
XMLModel
```

Reply:
Oh, those sound very idiomatic. Two thumbs up 👍👍